With `build -subset help` output "macro" subset components

BruceForstall opened the pull request:
Subsets like "clr", "mono", "host" are "macro" subset names that imply several other subset names. To try to explain how these relate to all the other names in `build -subset help`, actually output what they "expand" into.

e.g.,
```
  Accepted Subset values:
  - Clr
      The full CoreCLR runtime. Equivalent to: clr.native+linuxdac+clr.corelib+clr.tools+clr.nativecorelib+clr.packages+clr.nativeaotlibs
  - Clr.NativePrereqs
      Managed tools that support building the native components of the runtime (such as DacTableGen).
  - Clr.ILTools
      The CoreCLR IL tools (ilasm/ildasm).
  - Clr.Runtime
      The CoreCLR .NET runtime. Includes clr.jit, clr.iltools, clr.hosts.
  - Clr.Native
      All CoreCLR native non-test components, including the runtime, jits, and other native tools. Includes clr.hosts, clr.runtime, clr.jit, clr.alljits, clr.paltests, clr.iltools, clr.nativeaotruntime, clr.spmi.
  - Clr.NativeAotLibs
      The CoreCLR native AOT CoreLib and other low level class libraries.
  - Clr.NativeAotRuntime
      The stripped-down CoreCLR native AOT runtime.
  - Clr.PalTests [only runs on demand]
      The CoreCLR PAL tests.
  - Clr.PalTestList [only runs on demand]
      Generate the list of the CoreCLR PAL tests. When using the command line, use Clr.PalTests instead.
  - Clr.Hosts
      The CoreCLR corerun test host.
  - Clr.Jit
      The JIT for the CoreCLR .NET runtime.
  - Clr.AllJits
      All of the cross-targeting JIT compilers for the CoreCLR .NET runtime.
  - Clr.Spmi
      SuperPMI, a tool for CoreCLR JIT testing.
  - Clr.CoreLib
      The managed System.Private.CoreLib library for CoreCLR.
  - Clr.NativeCoreLib
      Run crossgen on System.Private.CoreLib library for CoreCLR.
  - Clr.Tools
      Managed tools that support CoreCLR development and testing.
  - Clr.Packages
      The projects that produce NuGet packages for the CoreCLR runtime, crossgen, and IL tools.
  - LinuxDac
      The cross-OS Windows->libc-based Linux DAC. Skipped on x86.
  - AlpineDac [only runs on demand]
      The cross-OS Windows->musl-libc-based Linux DAC. Skipped on x86
  - CrossDacPack [only runs on demand]
      Packaging of cross OS DAC. Requires all assets needed to be present at a folder specified by . See 'Microsoft.CrossOsDiag.Private.CoreCLR.proj' for details.
  - Mono
      The Mono runtime and CoreLib. Equivalent to: mono.runtime+mono.corelib+mono.packages
  - Mono.Runtime
      The Mono .NET runtime.
  - Mono.AotCross
      The cross-compiler runtime for Mono AOT.
  - Mono.CoreLib
      The managed System.Private.CoreLib library for Mono.
  - Mono.Packages
      The projects that produce NuGet packages for the Mono runtime.
  - Mono.WasmRuntime
      The WebAssembly runtime.
  - Mono.MsCorDbi
      The implementation of ICorDebug interface.
  - Mono.Workloads [only runs on demand]
      Builds the installers and the insertion metadata for Blazor workloads.
  - Host
      The .NET hosts, packages, hosting libraries, and tests. Equivalent to: host.native+host.tools+host.pkg+host.tests
  - Host.Native
      The .NET hosts.
  - Host.Pkg
      The .NET host packages.
  - Host.Tools
      The .NET hosting libraries.
  - Host.Tests
      The .NET hosting tests.
  - Libs
      The libraries native part, refs and source assemblies, test infra and packages, but NOT the tests (use Libs.Tests to request those explicitly). Equivalent to: libs.native+libs.ref+libs.src+libs.pretest
  - Libs.Native
      The native libraries used in the shared framework.
  - Libs.Ref
      The managed reference libraries.
  - Libs.Src
      The managed implementation libraries.
  - Libs.PreTest
      Test assets which are necessary to run tests.
  - Libs.Packages
      The projects that produce NuGet packages from libraries.
  - Libs.Tests [only runs on demand]
      The test projects. Note that building this doesn't execute tests: you must also pass the '-test' argument.
  - Packs
      Builds the shared framework packs, archives, bundles, installers, and the framework pack tests. Equivalent to: packs.product+packs.tests
  - Packs.Product
      Builds the shared framework packs, archives, bundles, and installers.
  - Packs.Installers
      Builds the shared framework bundles and installers.
  - Packs.Tests
      The framework pack tests.
  - publish [only runs on demand]
      Generate asset manifests and prepare to publish to BAR.
  - RegenerateDownloadTable [only runs on demand]
      Regenerates the nightly build download table
```

BruceForstall committed:
With `build -subset help` output "macro" subset components

BruceForstall commented:
Is this actually correct? Are there other "macros" that I missed? Does anyone else think this is helpful?

@jkoritzinsky @MichalStrehovsky @hoyosjs @safern

jkoritzinsky reviewed (commented):
As part of this, we should also update any subsets that are not included in these subsets to have the `OnDemand="true"` metadata.

hoyosjs commented on eng/Subsets.props:
```diff
@@ -101,7 +101,7 @@
   <ItemGroup>
     <!-- CoreClr -->
-    <SubsetName Include="Clr" Description="The CoreCLR runtime, LinuxDac, CoreLib (+ native), tools and packages." />
+    <SubsetName Include="Clr" Description="The full CoreCLR runtime: $(DefaultCoreClrSubsets)" />
```
@MichalStrehovsky @jkotas Does it make sense that `build.cmd clr` builds nativeaotlibs but not nativeaotruntime?

MichalStrehovsky replied:
Clr includes clr.native and that's all native code under src/coreclr including clr.nativeaotruntime, clr.iltools, etc.

I just realized I put `clr.spmi` into the `DefaultCoreClrSubsets` but that one is also redundant with `clr.native`. We could remove it; it's a NOP.

hoyosjs commented on eng/Subsets.props:
```diff
@@ -101,7 +101,7 @@
+    <SubsetName Include="Clr" Description="The full CoreCLR runtime: $(DefaultCoreClrSubsets)" />
     <SubsetName Include="Clr.NativePrereqs" Description="Managed tools that support building the native components of the runtime (such as DacTableGen)." />
     <SubsetName Include="Clr.ILTools" Description="The CoreCLR IL tools." />
     <SubsetName Include="Clr.Runtime" Description="The CoreCLR .NET runtime." />
```
@BruceForstall do you feel it makes sense that the default coreclr set includes SPMI?

BruceForstall replied:
It's probably not required; we'd just need to make sure all the JIT automation that requires it does actually build it.

BruceForstall replied:
> we'd just need to make sure all the JIT automation that requires it does actually build it.

Although _maybe_ https://github.com/dotnet/runtime/pull/64035 already did everything that is necessary w.r.t. CI changes required.

BruceForstall replied:
@dotnet/jit-contrib There's a proposal here to remove "clr.spmi" (the build of the SuperPMI tools) from the default "clr" component build.

(I'm not sure how the "clr.native" component works. It looks like if you use "clr.native" then src\coreclr\runtime.proj calls build-runtime.cmd/sh with no "-component" arguments, so the `__CMakeTarget` ends up being `install`, which builds everything? So using "clr.native" will still build everything? But "clr" uses "clr.native", so not sure how we actually "remove" clr.spmi from the "clr" component.)

hoyosjs replied:
We'd have to make clr.native in the default be clr.runtime+clr.iltools basically. I was just thinking that most people rarely use SPMI. Not sure libs dev should learn the long incantation.

jkoritzinsky replied:
I think @hoyosjs's question is: Should `clr.runtime` include the `clr.spmi` subset? I think having `clr` build all of the native code is still the correct thing.

MichalStrehovsky replied:
One could make the same argument about whether the `clr` should include the Linux DAC. I've never in my life used it and it takes longer to build than spmi. We also build the JIT 6 times because crossgen2/NativeAOT want to be able to crosstarget, but most people do not crosstarget in their day-to-day (crossgen2 devs included).

I think most devs who do daily work in the clr tree already have their own invocations of the build script and don't build all of `clr`. The `clr` subset is good for people new to the project that don't know what they want. So they just get everything.

hoyosjs replied:
@dotnet/runtime-infrastructure Do we see `build clr` as the product build, or as the inner loop? yml can easily have the longer invocation as needed, but I feel the short one should really be catering to the common case scenario (runtime + corelib) + crossgening maybe even optional. Probably shouldn't be part of this PR, but that could enable an easier workflow for a lot of people

BruceForstall replied:
I like the model where `build clr` is the "easy; build everything (related to coreclr)" model, and as you get more sophisticated, you should be able to only build what you care about by using some longer incantation.

BruceForstall committed:
Update text

Remove `clr.spmi` from default list; it's already there due to `clr.native`.

BruceForstall committed:
Clarify alljits

BruceForstall commented:
> As part of this, we should also update any subsets that are not included in these subsets to have the `OnDemand="true"` metadata.

I couldn't find any (in CLR) that were not correctly marked (by my understanding).

BruceForstall commented:
All: I updated the help text, and added documentation text for clr.native and clr.runtime, which include other subsets due to how the build-runtime script (and runtime.proj) are implemented.

This is ready for review.

hoyosjs reviewed (approved):
LGTM from the coreclr perspective. Also, OnDemand seems up to date for the src/coreclr components.
